### PR TITLE
gh-108511: Add C API functions which do not silently ignore errors

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Add the following functions:

* PyObject_HasAttrWithError()
* PyObject_HasAttrStringWithError()
* PyMapping_HasKeyWithError()
* PyMapping_HasKeyStringWithError()

* Issue: gh-108511

----
:books: Documentation preview :books:: https://cpython-previews--109025.org.readthedocs.build/

### CODE REVIEW

**Review by methane — APPROVED** [review]

**Comment by vstinner** [comment]
I wrote a first PR to add PyMapping_HasKeyWithError() and PyMapping_HasKeyStringWithError() functions to pythoncapi-compat: https://github.com/python/pythoncapi-compat/pull/73

File: Doc/c-api/mapping.rst [hunk 1]
```diff
@@ -76,6 +76,24 @@ See also :c:func:`PyObject_GetItem`, :c:func:`PyObject_SetItem` and
    rather than a :c:expr:`PyObject*`.
 
 
+.. c:function:: int PyMapping_HasKeyWithError(PyObject *o, PyObject *key)
+
+   Return ``1`` if the mapping object has the key *key* and ``0`` otherwise.
+   This is equivalent to the Python expression ``key in o``.
+   On failure, return ``-1``.
```
Comment (scoder) [inline Doc/c-api/mapping.rst:83]: Given the relevance of error handling for these new functions, I think the docs should mention explicitly under which conditions users must expect exceptions.
```suggestion
   On failure, return ``-1`` with an exception set.
```

File: Doc/c-api/object.rst [hunk 2]
```diff
@@ -27,6 +27,24 @@ Object Protocol
    instead of the :func:`repr`.
 
 
+.. c:function:: int PyObject_HasAttrWithError(PyObject *o, const char *attr_name)
+
+   Returns ``1`` if *o* has the attribute *attr_name*, and ``0`` otherwise.
+   This is equivalent to the Python expression ``hasattr(o, attr_name)``.
+   On failure, return ``-1``.
```
Comment (scoder) [inline Doc/c-api/object.rst:34]: Given the relevance of error handling for these new functions, I think the docs should mention explicitly under which conditions users must expect exceptions.
```suggestion
   On failure, return ``-1`` with an exception set.
```

File: Doc/c-api/mapping.rst (hunk 1 above)
Comment (vstinner) [reply Doc/c-api/mapping.rst:83]: Sometimes I write "On error, raise an exception and return -1." Even if in C, technically, it's more "setting" an exception, I like to use Python "raise" semantics in C, it's easy for me to map C/Python this way :-)

**Pull request merged**